File: source/modulo_core/test/python/test_joint_positions.py
import rclpy
from modulo_core import JointPositionsBroadcaster, JointPositionsListener
from modulo_interfaces.msg import JointPositions


def test_send_lookup(ros_context):
    node = rclpy.create_node("test_joint_positions")

    listener = JointPositionsListener(node)
    broadcaster = JointPositionsBroadcaster(node)

    joint_pos = JointPositions()
    joint_pos.header.frame_id = "foo"
    joint_pos.joint_names = ["one", "two"]
    joint_pos.positions = [1.0, 2.0]
    broadcaster.send(joint_pos)

    rclpy.spin_once(node)
    
    joint_positions = listener.lookup("foo")
    assert joint_positions.header.frame_id == joint_pos.header.frame_id
    assert joint_positions.joint_names == joint_pos.joint_names
    assert joint_positions.positions == joint_pos.positions

File: source/modulo_core/test/cpp/test_joint_positions.cpp
#include <gtest/gtest.h>

#include "modulo_core/JointPositionsBroadcaster.hpp"
#include "modulo_core/JointPositionsListener.hpp"

using namespace modulo_core;

TEST(TestJointPositions, SendLookup) {
  rclcpp::init(0, nullptr);

  auto node = rclcpp::Node::make_shared("test_joint_positions");

  JointPositionsListener listener(node);
  JointPositionsBroadcaster broadcaster(node);

  modulo_interfaces::msg::JointPositions joint_pos;
  joint_pos.header.frame_id = "foo";
  joint_pos.joint_names = {"one", "two"};
  joint_pos.positions = {1.0, 2.0};
  broadcaster.send(joint_pos);

  rclcpp::spin_some(node);
  rclcpp::sleep_for(std::chrono::milliseconds(10));

  state_representation::JointPositions joint_positions;
  EXPECT_NO_THROW(joint_positions = listener.lookup<state_representation::JointPositions>("foo"));
  EXPECT_EQ(joint_positions.get_name(), joint_pos.header.frame_id);
  EXPECT_EQ(joint_positions.get_names(), joint_pos.joint_names);
  EXPECT_EQ(joint_positions.to_std_vector(), joint_pos.positions);
}
